### PR TITLE
Cortex-M33 Instruction Tracing Without a Debugger

### PR DESCRIPTION
**Pull request by chrisc11**

### CODE REVIEW

File: _drafts/instruction-tracing-mtb-m33.md [hunk 1]
```diff
@@ -0,0 +1,752 @@
+---
+title: "Cortex-M33 Instruction Tracing Without a Debugger"
+description: "A step by step guide on how to configure and leverage the Cortex M33 Micro Trace Buffer (MTB) for instruction tracing using Dialog Semiconductor's DA14695 MCU"
+tag: [cortex-m]
+author: chris
+---
+
+In a previous post, we have talked about debugging memory corruption issues by making use of [watchpoints]({% post_url 2020-09-16-cortex-m-watchpoints %}). You may recall in that post we had to _reproduce_ the failure so we could halt the core with a watchpoint installed and debug when it happened.
```
Comment (tyhoff) [inline _drafts/instruction-tracing-mtb-m33.md:8]:
```suggestion
In a previous post, we talked about debugging memory corruption issues by making use of [watchpoints]({% post_url 2020-09-16-cortex-m-watchpoints %}). You may recall in that post we had to _reproduce_ the failure so we could halt the core with a watchpoint installed and debug when it happened.
```

File: _drafts/instruction-tracing-mtb-m33.md [hunk 2]
```diff
@@ -0,0 +1,752 @@
...
+# Flash & Debug
+$ make flash
+(gdb)
+```
+
+> Note the example App loads code directly into RAM so you will have to re-flash the application using `make flash` after running a test case.
+
+### Basic Usage Example
+
+The default mode for the test application does nothing exciting but will serve as a good way to sanity check our decoder is working as expected. When the application start's up it enters an infinite loop. The loop halts with a breakpoint instruction every five iterations:
```
Comment (tyhoff) [inline _drafts/instruction-tracing-mtb-m33.md:352]:
```suggestion
The default mode for the test application does nothing exciting but will serve as a good way to sanity check our decoder is working as expected. When the application starts up, it enters an infinite loop. The loop halts with a breakpoint instruction every five iterations:
```

File: _drafts/instruction-tracing-mtb-m33.md [hunk 3]
```diff
@@ -0,0 +1,752 @@
...
+```
+
+Neat, what we are looking at here is a trace of the entire instruction flow up to the breakpoint.
+
+If we want to view the exact instructions executed we can take the destination address from the previous packet and the source from the next packet and dissasemble the range with gdb. We see the `0x20000348` we identified above has been recorded in the trace.
+
+For example for the first `D` -> `S` transition we have above:
+
+```
+(gdb) disassemble 0x20000368,0x2000036e+4
```
Comment (tyhoff) [inline _drafts/instruction-tracing-mtb-m33.md:445]: MABYE: Use `/s` argument to print the source code inline

File: _drafts/instruction-tracing-mtb-m33.md [hunk 4]
```diff
@@ -0,0 +1,752 @@
...
+   0x20000344 <infinite_loop+32>:	subs	r2, r1, r3
+   0x20000346 <infinite_loop+34>:	cmp	r2, #0
+   0x20000348 <infinite_loop+36>:	bne.n	0x2000032a <infinite_loop+6>
+```
+
+Neat, so two trace packets (16 bytes of data) tells us the exact last 23 instructions which were executed by the processor! Both ranges examined end with branch instructions as expected. Our decoder appears to be working as expected!
+
+### Stack Overflow Debug Example
+
+#### Context
```
Comment (tyhoff) [inline _drafts/instruction-tracing-mtb-m33.md:482]:
```suggestion
### Stack Overflow Debug Example

Let's now go over an example of how to debug a stack overflow using the MTB.

#### Context
```

File: _drafts/instruction-tracing-mtb-m33.md [hunk 5]
```diff
@@ -0,0 +1,752 @@
...
+    /* Avoid dependency on memset() if it is not required. */
+    #if ( tskSET_NEW_STACKS_TO_KNOWN_VALUE == 1 )
+        {
+            /* Fill the stack with a known value to assist debugging. */
+            ( void ) memset( pxNewTCB->pxStack, ( int ) tskSTACK_FILL_BYTE, ( size_t ) ulStackDepth * sizeof( StackType_t ) );
+        }
+    #endif /* tskSET_NEW_STACKS_TO_KNOWN_VALUE */
+```
+
+Then, everytime a task context switch takes place the bottom of the stack is checked to see if it matches the known pattern. The FreeRTOS logic looks like this:
```
Comment (tyhoff) [inline _drafts/instruction-tracing-mtb-m33.md:498]:
```suggestion
Then, every time a task context switch takes place, the bottom of the stack is checked to see if it matches the known pattern. The FreeRTOS logic looks like this:
```

File: _drafts/instruction-tracing-mtb-m33.md [hunk 6]
```diff
@@ -0,0 +1,752 @@
...
+static void bad_asm_function(void) {
+  __asm("mov r0, 0 \n"
+        "ldr r2, =3204505165 \n"
+        "mov lr, 0 \n"
+        "bx r2 \n"
+        );
+}
+```
+
+But hopefullt, the example conveys the general idea how this type of trace can be useful for root causing this class of bug!
```
Comment (tyhoff) [inline _drafts/instruction-tracing-mtb-m33.md:718]:
```suggestion
Although contrived, I hope the example conveys the general idea how this type of trace can be useful for root causing this class of bug!
```

**Review by tyhoff — APPROVED** [review]
Awesome post! Loved it. Simple and sweet.

**Comment by franc0is** [comment]
This is great. Ship it!

**Pull request merged**